### PR TITLE
Assorted satchel and ore-scoop bugfixes and QOL changes

### LINKED ISSUE #7614: Stacked ore counts as one in the mining satchel
### Labels

[BUG]

### Description

When stacked or is put into a mining satchel it only adds 1 to the contained item count

### Steps to reproduce

1. Stack some ore
2. Place ore into mining satchel
3. Observe how the counter only goes up by 1

### Expected Behavior

Placing items into the satchel should unstack the ore

### Additional Information & Screenshots

_No response_

### LINKED ISSUE #12900: Plant satchels keep the number of items displayed after emptying them into vendors.
### Labels

[BUG]

### Description

Plant satchels display the number of items that were in it after they are emptied into player vendors, despite being empty.

### Steps to reproduce

1. Put stuff in produce satchel.
2. Click-drag onto player vendor to stock it.
3. Look at satchel after and see it still has the number there despite being empty.

### Expected Behavior

For the display to return to 0 once emptied into the vendor

### Additional Information & Screenshots

_No response_

### PR DESCRIPTION
**Pull request by M-Earthfire**
[QOL][Game-Objects]

## About the PR 

This PR introduces various changes to the behaviour of produce/mining satchels and connected interactions, with a focus on the mining satchel and issues connected to it.

- Satchels can now be assigned exceptions within `var/list/exceptions` for what items within `var/list/allowed` they should not use. Also, satchels can now be modified with `var/maximal_w_class` to constrain which is the biggest weight class a possible item should have.
- When trying to put an item with multiple stacks in a satchel (e.g. stacked ores) the stack gets split up automatically and single ores inserted into the satchel. A safety limit of 20 splits per interaction is added so a single interaction should rather not be able to create 100 items. `proc/split_stack_into_satchel` was added to quickly split the stack into a satchel. This fixes #7614
- This behaviour was added to ore scoops and pod ore scoops as well
- `proc/check_valid_content` was added to quickly check if a satchel can contain a certain item that is passed as argument.
- When harvesting items from plants with a satchel, only items are added into the satchel that can actually be added into the satchel
- Most interactions i found that dump items from satchels both set `tooltip_rebuild` to 1 and call `UpdateIcon()` on the satchel. This fixes #12900
- Ore scoops now don't need two hands to be able to switch satchels. Clicking with an ore scoop on a mining satchel now switches the ore scoops and places the old satchel in the container the former was. If it wasn't in a container, it tries to place the ejected satchel in a free hand, if able. This mimics the behaviour of using kinetic guns on ammo to quick reload them.

## Why's this needed? 

Firstly, this PR prevents stacking ores in a bunch so you cannot place 1000+ ores in a satchel anymore. 

Beyond that, by changing how ore scoop works, it enables much more intuitive handling or ore scoops for miners without accidently dropping of satchels... to be honest, that was grinding my gears to a degree it made me make this PR

Lastly, this standartises behaviour of satchels accross the board and makes them more versatile and easy to work with.

## Changelog 

```changelog
(u)Lord_Earthfire
(+)Satchels (e.g. mining, produce satchels) cannot hold stacks of items anymore. Stuffing a stack of items splits part of it up into singular items.
(+)The ore scoop doesn't need two hands to be reloaded anymore. Enjoy quick reloading of your scoop without annoying dropping of mining satchels!
```

### CODE REVIEW

**Comment by CheffieGithub** [comment]
big balance change with the stacking limits, probably mention that in the title

**Comment by M-Earthfire** [comment]
> big balance change with the stacking limits, probably mention that in the title

That's why i put it explicitely in the changelog. It should be known about,  although it mostly affects people trying to powergame satchels, so i don't really think it's a big balance change.

**Comment by CheffieGithub** [comment]
Its been around for so long it may as well be a balance change but its a good change

**Review by virvatuli-gh — APPROVED** [review]
I can't comment on the code of this but I like the proposed changes quite a bit. The ore scoop QoL in particular is something that's been needed for a long time (it's bugged me forever when playing miner) and the super-stacking of items in bags is a nonsensical bug that should have been fixed ages ago before it became considered a "feature". :thumbsup:

File: code/modules/chemistry/Chemistry-Machinery.dm [hunk 1]
```diff
@@ -1267,6 +1267,7 @@ TYPEINFO(/obj/machinery/chemicompiler_stationary)
 				boutput(user, "<span class='notice'>[loadcount] items were loaded from the satchel!</span>")
 
 			S.UpdateIcon()
+			S:tooltip_rebuild = 1
```
Comment (ZeWaka) [inline code/modules/chemistry/Chemistry-Machinery.dm:1270]: does this need to be a `:`?

File: code/obj/item/material.dm [hunk 2]
```diff
@@ -45,16 +45,31 @@
...
-				var/oreamt = length(W.contents)
-				boutput(user, "<span class='notice'>You put [src] in [W].</span>")
-				src.desc = "A leather bag. It holds [oreamt]/[W:maxitems] [W:itemstring]."
-				if (oreamt == W:maxitems) boutput(user, "<span class='notice'>[W] is now full!</span>")
-				W:UpdateIcon()
+			var/obj/item/satchel/mining/manipulated_satchel = W
+			if (!manipulated_satchel.check_valid_content(src))
+				boutput(user, "<span class='alert'>[manipulated_satchel] cannot hold that kind of item!</span>")
+				return
+			if (manipulated_satchel.contents.len < manipulated_satchel.maxitems)
```
Comment (ZeWaka) [inline code/obj/item/material.dm:52]: length()

File: code/obj/item/material.dm [hunk 3]
```diff
@@ -45,16 +45,31 @@
...
+					var/amount_of_stack_splits = manipulated_satchel.split_stack_into_satchel(src, user)
+					if (amount_of_stack_splits == manipulated_satchel.max_stack_scoop)
+						max_stack_reached = TRUE
+				else
+					boutput(user, "<span class='notice'>You put [src] in [manipulated_satchel].</span>")
+				if (!max_stack_reached && (manipulated_satchel.contents.len < manipulated_satchel.maxitems)) // if we split up the item and it was more than the satchel can find we should not add the rest
+					user.u_equip(src)
+					src.set_loc(manipulated_satchel)
+					src.dropped(user)
+				if (manipulated_satchel.contents.len == manipulated_satchel.maxitems)
```
Comment (ZeWaka) [inline code/obj/item/material.dm:65]: ditto

File: code/obj/item/material.dm [hunk 4]
```diff
@@ -45,16 +45,31 @@
...
+			if (manipulated_satchel.contents.len < manipulated_satchel.maxitems)
+				var/max_stack_reached = FALSE
+				if (src.amount > 1)
+					boutput(user, "<span class='notice'>You begin to fill [manipulated_satchel] with [src].</span>")
+					var/amount_of_stack_splits = manipulated_satchel.split_stack_into_satchel(src, user)
+					if (amount_of_stack_splits == manipulated_satchel.max_stack_scoop)
+						max_stack_reached = TRUE
+				else
+					boutput(user, "<span class='notice'>You put [src] in [manipulated_satchel].</span>")
+				if (!max_stack_reached && (manipulated_satchel.contents.len < manipulated_satchel.maxitems)) // if we split up the item and it was more than the satchel can find we should not add the rest
```
Comment (ZeWaka) [inline code/obj/item/material.dm:61]: ditto

File: code/obj/item/material.dm (hunk 3 above)
Comment (ZeWaka) [inline code/obj/item/material.dm:65]: ditto

File: code/obj/item/material.dm [hunk 5]
```diff
@@ -75,8 +90,19 @@
...
 			var/mob/living/H = AM
 			var/obj/item/ore_scoop/S = H.get_equipped_ore_scoop()
-			if (S?.satchel && length(S.satchel.contents) < S.satchel.maxitems && src.scoopable)
-				src.set_loc(S.satchel)
+			if (S?.satchel && length(S.satchel.contents) < S.satchel.maxitems && src.scoopable && S.satchel.check_valid_content(src))
+				var/max_stack_reached = FALSE
+				if (src.amount > 1)
+					var/increment = 0
+					//since we need to add additional manipulation to the item in hand, we won't touch the last item here
+					var/amount_of_stack_splits = min(S.satchel.maxitems - S.satchel.contents.len, src.amount - 1, S.satchel.max_stack_scoop)
```
Comment (ZeWaka) [inline code/obj/item/material.dm:98]: ditto

File: code/obj/item/material.dm [hunk 6]
```diff
@@ -75,8 +90,19 @@
...
+					if (amount_of_stack_splits == S.satchel.max_stack_scoop)
+						max_stack_reached = TRUE
+					for (increment = 0, increment < amount_of_stack_splits, increment++)
+						var/obj/item/splitted_stack = src.split_stack(1)
+						splitted_stack.set_loc(S.satchel)
+				if (!max_stack_reached && (S.satchel.contents.len < S.satchel.maxitems))
```
Comment (ZeWaka) [inline code/obj/item/material.dm:104]: ditto

File: code/obj/item/material.dm [hunk 7]
```diff
@@ -87,7 +113,18 @@
...
+				if (src.amount > 1)
+					var/increment = 0
+					//since we need to add additional manipulation to the item in hand, we won't touch the last item here
+					var/amount_of_stack_splits = min(SCOOP.capacity - SCOOP.contents.len, src.amount - 1, SCOOP.max_stack_scoop)
+					if (amount_of_stack_splits == SCOOP.max_stack_scoop)
+						max_stack_reached = TRUE
+					for (increment = 0, increment < amount_of_stack_splits, increment++)
+						var/obj/item/splitted_stack = src.split_stack(1)
+						splitted_stack.set_loc(SCOOP)
+				if (!max_stack_reached && (SCOOP.contents.len < SCOOP.capacity)) // if we split up the item and it was more than the satchel can find we should not add the rest
```
Comment (ZeWaka) [inline code/obj/item/material.dm:126]: ditto

File: code/obj/item/material.dm [hunk 8]
```diff
@@ -87,7 +113,18 @@
 				var/obj/item/shipcomponent/secondary_system/orescoop/SCOOP = V.sec_system
 				if (SCOOP.contents.len >= SCOOP.capacity || !src.scoopable)
 					return
-				src.set_loc(SCOOP)
+				var/max_stack_reached = FALSE
+				if (src.amount > 1)
+					var/increment = 0
+					//since we need to add additional manipulation to the item in hand, we won't touch the last item here
+					var/amount_of_stack_splits = min(SCOOP.capacity - SCOOP.contents.len, src.amount - 1, SCOOP.max_stack_scoop)
```
Comment (ZeWaka) [inline code/obj/item/material.dm:120]: ditto

File: code/obj/item/satchel.dm [hunk 9]
```diff
@@ -8,7 +8,10 @@
 	w_class = W_CLASS_TINY
 	event_handler_flags = USE_FLUID_ENTER | NO_MOUSEDROP_QOL
 	var/maxitems = 50
+	var/max_stack_scoop = 20 //! if you try to put stacks inside the item, this one limits how much you can in one action. Creating 100 items out of a stack in a single action should not happen.
 	var/list/allowed = list(/obj/item/)
+	var/list/exceptions = list() //! this list are for items that are in the allowed-list for other reasons, but should not be able to be put in satchels
```
Comment (ZeWaka) [inline code/obj/item/satchel.dm:13]: please initialize these in new()
https://hackmd.io/@goonstation/code#StartupRuntime-trade-offs-with-lists-and-the-%E2%80%9Chidden%E2%80%9D-init-proc

File: code/obj/item/satchel.dm [hunk 10]
```diff
@@ -149,6 +170,40 @@
...
+		// This proc splits an object with multiple stacks and stuff it into the satchel until either
+		// The satchel is full
+		// all but the origin item of the stack is in the satchel
+		// the safety-amount of items were stuffed to prevent laggs.
+		// The proc returns the amount of times splits were created and stuffed
+		if (!(item_to_split) || (item_to_split.amount <= 1))
+			return 0
+		var/increment = 0
+		//since we need to add additional manipulation to the item in hand, we won't touch the last item here
+		var/amount_of_stack_splits = min(src.maxitems - src.contents.len, item_to_split.amount - 1, src.max_stack_scoop)
```
Comment (ZeWaka) [inline code/obj/item/satchel.dm:183]: length()

**Review by ZeWaka — CHANGES_REQUESTED** [review]

File: code/datums/components/transfers.dm [hunk 11]
```diff
@@ -158,6 +158,8 @@
 				SEND_SIGNAL(parent, COMSIG_TRANSFER_INCOMING, I)
 				transfers++
 			incoming.UpdateIcon()
+			if (istype(incoming, /obj/item/satchel))
+				incoming:tooltip_rebuild = 1
```
[the remark this answers is not in the document]
Comment (M-Earthfire) [reply code/datums/components/transfers.dm:162]: You mean typing the tooltip of the satchel?

File: code/obj/item/satchel.dm [hunk 12]
```diff
@@ -8,35 +8,46 @@
 	w_class = W_CLASS_TINY
 	event_handler_flags = USE_FLUID_ENTER | NO_MOUSEDROP_QOL
 	var/maxitems = 50
-	var/list/allowed = list(/obj/item/)
+	var/max_stack_scoop = 20 //! if you try to put stacks inside the item, this one limits how much you can in one action. Creating 100 items out of a stack in a single action should not happen.
+	var/list/allowed
+	var/list/exceptions  //! this list are for items that are in the allowed-list for other reasons, but should not be able to be put in satchels
```
Comment (ZeWaka) [inline code/obj/item/satchel.dm:13]:
```suggestion
	var/list/allowed = null
	var/list/exceptions = null //! this list are for items that are in the allowed-list for other reasons, but should not be able to be put in satchels
```

**Review by ZeWaka — APPROVED** [review]

**Pull request merged**